src/arch/io/network.cc

```diff
@@ -621,6 +621,424 @@ void linux_tcp_conn_t::on_event(int /* events */) {
     event_watcher->stop_watching_for_errors();
 }
 
+/* This is the client version of the constructor. The base class constructor
+will establish a TCP connection to the peer at the given host:port and then we
+wrap the tcp connection in TLS using the configuration in the given tls_ctx. */
+linux_secure_tcp_conn_t::linux_secure_tcp_conn_t(SSL_CTX *tls_ctx, const ip_address_t &host, int port, signal_t *interruptor, int local_port) THROWS_ONLY(connect_failed_exc_t, interrupted_exc_t) :
+        linux_tcp_conn_t(host, port, interruptor, local_port) {
+    establish_conn(tls_ctx, SSL_connect);
+}
+
+/* This is the server version of the constructor */
+linux_secure_tcp_conn_t::linux_secure_tcp_conn_t(SSL_CTX *tls_ctx, fd_t sock) :
+        linux_tcp_conn_t(sock) {
+    try {
+        establish_conn(tls_ctx, SSL_accept);
+    } catch (const tcp_conn_t::connect_failed_exc_t &err) {
+        /* Ignore */
+        logNTC("linux_secure_tcp_conn_t server constructor exception: %s\n", err.info.c_str());
+    }
+}
+
+linux_secure_tcp_conn_t::~linux_secure_tcp_conn_t() THROWS_NOTHING {
+    assert_thread();
+
+    logNTC("linux_secure_tcp_conn_t destructor\n");
+
+    if (is_open()) shutdown();
+}
+
+/* Creates a new OpenSSL connection using the underlying fd and the given
+OpenSSL context. The given handshake function is used to perform the handshake
+for the client or server using SSL_connect or SSL_accept respectively. */
+void linux_secure_tcp_conn_t::establish_conn(SSL_CTX *tls_ctx, int (*handshake)(SSL *) ) THROWS_ONLY(connect_failed_exc_t) {
+    conn = SSL_new(tls_ctx);
```

Comment:
I've moved this allocation into the constructors. Since only one raises an exception, I've added a try/catch block which frees the allocated memory before re-throwing the exception.